File: src/main/java/hzt/aoc/day13/Part2ShuttleSearch.java
package hzt.aoc.day13;

import hzt.collections.ArrayX;
import hzt.collections.ListX;
import hzt.collections.MutableListX;
import hzt.utils.Transformable;

import java.util.List;

public class Part2ShuttleSearch extends Day13Challenge {

    public Part2ShuttleSearch() {
        super("part 2",
                "What is the earliest timestamp such that all of the listed bus IDs depart at " +
                        "offsets matching their positions in the list?");
    }

    @Override
    protected String solve(List<String> lines) {
        return ArrayX.of(lines.get(1).split(","))
                .map(x -> "x".equals(x) ? "-1" : x)
                .map(Integer::parseInt)
                .run(this::second)
                .let(String::valueOf);
    }

    // Chinese remainder theorem implementation
    private long second(ListX<Integer> busIdsWithBlanks) {
        var equations = MutableListX.<Equation>empty();
        for (int i = 0; i < busIdsWithBlanks.size(); i++) {
            int busId = busIdsWithBlanks.get(i);
            if (busId != -1) {
                equations.add(new Equation(-i, busId));
            }
        }

        long moduloProduct = equations.fold(1L, (acc,  equation) -> acc * equation.modulo);

        for (Equation equation : equations) {
            equation.n = moduloProduct / equation.modulo;
            equation.w = euclideanGeneralTheorem(equation.n, equation.modulo);
            if (equation.w < 0) {
                equation.w += equation.modulo;
            }
        }
        return Transformable.from(equations.sumOfLongs(equation -> equation.minIndex * equation.n * equation.w))
                .let(solution -> properModulo(solution, moduloProduct));
    }

    private long euclideanGeneralTheorem(long a1, long b1) {
        long b = a1;
        long a = b1;

        long x = 0;
        long y = 1;
        long u = 1;
        long v = 0;
        while (b != 0) {
            long q = a / b;
            long r = properModulo(a, b);
            long m = x - u * q;
            long n = y - v * q;
            a = b;
            b = r;
            x = u;
            y = v;
            u = m;
            v = n;
        }
        if (a != 1) {
            LOGGER.error("!" + a);
        }
        return x;
    }

    private long properModulo(long a, long b) {
        b = Math.abs(b);
        if (a < 0) {
            return b - ((-a) % b);
        } else {
            return a % b;
        }
    }

    static class Equation {

        long minIndex;
        long modulo;
        long n;
        long w;

        public Equation(long minIndex, long modulo) {
            this.minIndex = minIndex;
            this.modulo = modulo;
        }
    }
}

File: src/main/kotlin/hzt/aoc/day13/Part2ShuttleSearch.kt
package hzt.aoc.day13

import kotlin.math.abs

class Part2ShuttleSearch : Day13Challenge(
    "part 2",
    "What is the earliest timestamp such that all of the listed bus IDs depart at " +
            "offsets matching their positions in the list"
) {
    override fun solve(inputList: List<String>): String {
        val busIdsWithBlanks = sequenceOf(*inputList[1].split(",".toRegex()).toTypedArray())
            .map { if (it == "x") "-1" else it }
            .map(String::toInt)
            .toList()
        return second(busIdsWithBlanks).toString()
    }

    // Chinese remainder theorem implementation
    private fun second(busIdsWithBlanks: List<Int>): Long {
        val equations: MutableList<Equation> = ArrayList()
        for (i in busIdsWithBlanks.indices) {
            val busId = busIdsWithBlanks[i]
            if (busId != -1) {
                equations.add(Equation((-i).toLong(), busId.toLong()))
            }
        }
        val moduloProduct = equations
            .map(Equation::modulo)
            .reduce { a, b -> a * b }

        for (equation in equations) {
            equation.n = moduloProduct / equation.modulo
            equation.w = euclideanGeneralTheorem(equation.n, equation.modulo)
            if (equation.w < 0) equation.w += equation.modulo
        }
        val solution = equations.asSequence()
            .map { it.minIndex * it.n * it.w }
            .sum()
        return properModulo(solution, moduloProduct)
    }

    private fun euclideanGeneralTheorem(a1: Long, b1: Long): Long {
        var b = a1
        var a = b1
        var x: Long = 0
        var y: Long = 1
        var u: Long = 1
        var v: Long = 0
        while (b != 0L) {
            val q = a / b
            val r = properModulo(a, b)
            val m = x - u * q
            val n = y - v * q
            a = b
            b = r
            x = u
            y = v
            u = m
            v = n
        }
        if (a != 1L) LOGGER.error("!$a")
        return x
    }

    private fun properModulo(a: Long, b: Long): Long {
        var bMut = b
        bMut = abs(bMut)
        return if (a < 0) bMut - -a % bMut else a % bMut
    }

    internal class Equation(var minIndex: Long, var modulo: Long) {
        var n: Long = 0
        var w: Long = 0
    }
}
